common/lanelet2_extension/lib/query.cpp

```diff
@@ -31,6 +31,135 @@ namespace lanelet
 {
 namespace utils
 {
+// Point
+void recurse (const lanelet::ConstPoint3d& prim, const lanelet::LaneletMapPtr ll_Map, query::direction check_dir, query::References& rfs)
+{
+  // no primitive lower than point, so always go back up : CHECK_PARENT
+  // get LineStrings that own this point
+  auto ls_list_owning_point = ll_Map->lineStringLayer.findUsages(prim);
+
+  // if it's not owned by anyone, do not record it since points are no meaningful objects in Lanelet
+  if (ls_list_owning_point.size() == 0)
+      return;
+
+  for (auto ls : ls_list_owning_point)
+  {
+    recurse(ls, ll_Map, query::direction::CHECK_PARENT, rfs);
+  }
+}
+
+// LineString
+void recurse (const lanelet::ConstLineString3d& prim, const lanelet::LaneletMapPtr ll_Map, query::direction check_dir, query::References& rfs)
+{
+  // go down in primitive layer
+  if (check_dir == query::CHECK_CHILD)
+  {
+    // loop through its child and call recurse down on it
+    for (auto p: prim)
+    {
+      recurse(p, ll_Map, check_dir, rfs);
+    }
+    // go back up once finished
+    return;
+  }
+  
+  // go up, CHECK_PARENT
+  // process lanelets owning this ls
+  auto llt_list_owning_ls = ll_Map->laneletLayer.findUsages(prim);
+
+  for (auto llt : llt_list_owning_ls)
+  {
+    // recurse up on this lanelet
+    recurse(llt, ll_Map, query::CHECK_PARENT, rfs);
+  }
+  
+  // similarly, process areas owning this ls
+  auto area_list_owning_ls = ll_Map->areaLayer.findUsages(prim);
+  for (auto area : area_list_owning_ls)
+  {
+    // recurse up on this area
+    recurse(area, ll_Map, query::CHECK_PARENT, rfs);
+  }
+
+  if (area_list_owning_ls.size() ==0 && llt_list_owning_ls.size() == 0)
+    rfs.lss.insert(prim);
+}
+
+// Lanelet
+void recurse (const lanelet::ConstLanelet& prim, const lanelet::LaneletMapPtr ll_Map, query::direction check_dir, query::References& rfs)
+{
+  // go down, query::CHECK_CHILD
+  if (check_dir == query::CHECK_CHILD)
+  {
+    // loop through its child and call recurse down on it
+    recurse(prim.leftBound(), ll_Map, check_dir, rfs);
+    recurse(prim.rightBound(), ll_Map, check_dir, rfs);
+    recurse(prim.centerline(), ll_Map, check_dir, rfs);
```

Comment:
Does the centerline ever become part of the linestring layer?

The centerline is not encoded in the map and is actually computed based on the bounds, so I'm not sure this call will ever have meaningful results. If you have found otherwise though we should keep it.